wgengine/tsdns/forwarder.go

```diff
@@ -285,3 +272,189 @@ func (f *forwarder) forward(query Packet) error {
 
 	return nil
 }
+
+// A fwdconn manages a single connection used to forward DNS requests.
+// Net link changes can cause a *net.UDPConn to become permanently unusable, particularly on macOS.
+// fwdconn detects such situations and transparently creates new connections.
+type fwdconn struct {
+	// logf allows a fwdconn to log
+	logf logger.Logf
+
+	// wg tracks the number of outstanding conn.Read and conn.Write calls
+	wg sync.WaitGroup
+	// change allows calls to read to block until a the network connection has been replaced
+	change *sync.Cond
+
+	// mu protects fields that follow it; it is is also change's Locker
+	mu sync.Mutex
+
+	// closed tracks whether fwdconn has been permanently closed
+	closed bool
+
+	// conn is the current active connection
+	conn *net.UDPConn
+}
+
+func newFwdConn(logf logger.Logf, idx int) *fwdconn {
+	c := new(fwdconn)
+	c.logf = logger.WithPrefix(logf, fmt.Sprintf("fwdconn %d: ", idx))
+	c.change = sync.NewCond(&c.mu)
+	// c.conn is created lazily in send
+	return c
+}
+
+// send sends packet to dst using fwdconn's connection.
+// It is best effort. It is UDP, after all. Failures are logged.
+func (c *fwdconn) send(packet []byte, dst net.Addr) {
+	b := backoff.NewBackoff("tsdns-fwdconn-send", c.logf, 30*time.Second)
```

Comment:
the common case is success. can you lazily allocate this?

Reply:
It doesn't move the benchmarks, but since it's done now, I'll leave it in place.